pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,338 @@
+limitations under the License.
+*/
+
+package container
+
+import (
+	"io"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+type PodSandboxID string
+
+// PodSandboxManager provides basic operations to create/delete and examine the
+// PodSandboxes.
+type PodSandboxManager interface {
+	// Create creates a sandbox based on the given config, and returns the
+	//  of the new sandbox.
+	Create(config *PodSandboxConfig) (PodSandboxID, error)
+	// Delete deletes the sandbox by its ID. If there are any running
+	// containers in the sandbox, they will be terminated as a side-effect.
+	Delete(id PodSandboxID) error
+	// List lists existing sandboxes, filtered by the given PodSandboxFilter.
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	// Status gets the status of the sandbox by ID.
+	Status(id PodSandboxID) (PodSandboxStatus, error)
+	// PortForward copies the data from the port in the PodSandBox to the
+	// stream.
+	// TODO: Should PortForward be in a separate interface.
+	PortForward(id PodSandboxID, port uint16, stream io.ReadWriteCloser) error
+}
+
+// PodSandboxConfig holds all the required and optional fields for creating a
+// sandbox.
+type PodSandboxConfig struct {
+	// Name is the name of the sandbox.
+	Name string
+	// Hostname is the hostname of the sandbox.
+	Hostname string
+	// DNSOptions sets the DNS options for the sandbox.
+	DNSOptions DNSOptions
+	// PortMappings lists the port mappings for the sandbox.
+	PortMappings []PortMapping
+	// Mounts lists the mounts to be added to the sandbox's filesystem.
+	Mounts []Mount
+	// Resources specifies the resource requirements for the sandbox.
+	// Note: On a Linux host, kubelet will create a pod-level cgroup and pass
+	// it as the cgroup parent for the PodSandbox. For some runtimes, this is
+	// sufficent to fulfill pod-level resource requirements, and this field
+	// will not be used. For others, e.g., hypervisor-based runtimes, explicit
+	// resource requirements for the sandbox are needed at creation time.
+	Resources PodSandboxResources
+	// Labels are key value pairs that may be used to scope and select individual resources.
+	Labels Labels
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+
+	// Linux contains configurations specific to Linux hosts.
+	Linux *LinuxPodSandboxConfig
+}
+
+// Labels are key value pairs that may be used to scope and select individual resources.
+// Label keys are of the form:
+//     label-key ::= prefixed-name | name
+//     prefixed-name ::= prefix '/' name
+//     prefix ::= DNS_SUBDOMAIN
+//     name ::= DNS_LABEL
+type Labels map[string]string
+
+// LinuxPodSandboxConfig holds platform-specific configuraions for Linux
+// host platforms and Linux-based containers.
+type LinuxPodSandboxConfig struct {
+	// CgroupParent is the parent cgroup of the sandbox.
+	CgroupParent string
```

Comment:
@derekwaynecarr Is it OK to run systemd agnostic runtimes on a machine running systemd? I'd hope the answer is yes. Kubelet can create a slice for the pod and then let the runtime use the slice cgroups for managing containers.

Kubelet can accept cgroup paths based on the driver. Not doing that will definitely affect usability. 

When interacting with runtimes, can we have the kubelet use cgroupfs path and let runtimes convert to an appropriate format based on their needs?